src/pcap/agent.go

```diff
@@ -0,0 +1,367 @@
+package pcap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/pcap"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+)
+
+// Agent is the central struct to which the handlers are attached.
+type Agent struct {
+	// done is used to gracefully shut down the agent, all ongoing streams terminate
+	// whenever this channel is closed.
+	done chan struct{}
+	// wg tracks any running streams.
+	// TODO: expose as metric?
+	wg sync.WaitGroup
+	// log carries the logger all session loggers are derived from.
+	log     *zap.Logger
+	bufConf BufferConf
+
+	UnimplementedAgentServer
+}
+
+// NewAgent creates a new ready-to-use agent. If the given logger is nil zap.L will
+// be used.
+func NewAgent(log *zap.Logger, bufConf BufferConf) (*Agent, error) {
+	if log == nil {
+		log = zap.L()
+	}
+	return &Agent{
+		done:    make(chan struct{}),
+		log:     log,
+		bufConf: bufConf,
+	}, nil
+}
+
+// Stop the server. This will gracefully stop any captures that are currently running
+// by closing Agent.done. Further calls to Stop have no effect.
+func (a *Agent) Stop() {
+	select {
+	case <-a.done:
+		// if the channel is already closed, we do nothing
+	default:
+		// otherwise the channel is still open and we close it
+		close(a.done)
+	}
+}
+
+// Wait for all open streams to terminate.
+func (a *Agent) Wait() {
+	a.wg.Wait()
+}
+
+// draining returns true after Agent.Stop has been called.
+func (a *Agent) draining() bool {
+	select {
+	case <-a.done:
+		// we only get here if the channel is closed since it is never written to
+		return true
+	default:
+		// channel is still open
+		return false
+	}
+}
+
+// Status handler for the pcap-agent. See AgentServer.Status documentation for details.
+func (a *Agent) Status(_ context.Context, _ *StatusRequest) (*StatusResponse, error) {
+	s := &StatusResponse{
+		CompatibilityLevel: CompatibilityLevel,
+		Health:             Health_UP,
+		Status:             "ok",
+	}
+
+	if a.draining() {
+		s.Health = Health_DRAINING
+		s.Status = "agent is shutting down"
+	}
+
+	return s, nil
+}
+
+// Capture handler for the pcap-agent. See AgentServer.Capture documentation for details.
+func (a *Agent) Capture(stream Agent_CaptureServer) (err error) {
+	a.wg.Add(1)
+	defer a.wg.Done()
+
+	log := a.log.With(zap.String("handler", "capture"))
+	defer func() {
+		if err != nil {
+			log.Error("capture ended unsuccessfully", zap.Error(err))
+		}
+	}()
+
+	ctx, cancel := WithCancelCause(stream.Context())
+	defer cancel(nil)
+
+	setVcapID(ctx, log)
```

Comment:
After looking into gRPC to see what we can do to implement middleware I'm afraid we won't be able to easily implement something like this. The interceptor (=gRPC middleware) doesn't allow for the context to be modified and is a generic implementation not tied to the underlying type of stream.

[StreamServerInterceptor](https://pkg.go.dev/google.golang.org/grpc#StreamServerInterceptor):
```go
type StreamServerInterceptor func(srv interface{}, ss ServerStream, info *StreamServerInfo, handler StreamHandler) error
```

The [UnaryServerInterceptor](https://pkg.go.dev/google.golang.org/grpc#UnaryServerInterceptor) on the other hand gets the context directly, can modify it, and pass it on. This would allow us to put anything we need to add to the logging into the context, but except for the status endpoints we always have streaming RPCs.

Maybe grpc/grpc-go#1805 would address this but it seems like it can take some time for this to be done.

Reply:
as discussed, ok for now. we will look into the middleware part later down the road.